common/sgx/eeid.c

```diff
@@ -255,8 +246,23 @@ oe_result_t oe_deserialize_eeid(
     OE_CHECK(deserialize_element(
         position,
         &remaining,
-        (uint8_t*)eeid->data,
-        eeid->data_size + eeid->signature_size));
+        (uint8_t*)&eeid->signature_size,
+        sizeof(eeid->signature_size)));
+    eeid = realloc(eeid, sizeof(oe_eeid_t) + eeid->signature_size);
+    if (!eeid)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+    OE_CHECK(deserialize_element(
+        position, &remaining, (uint8_t*)eeid->signature, eeid->signature_size));
+    OE_CHECK(deserialize_element(
+        position, &remaining, (uint8_t*)data_size, sizeof(*data_size)));
+    if (data)
+    {
+        *data = malloc(*data_size);
+        if (!*data)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+        OE_CHECK(deserialize_element(
+            position, &remaining, (uint8_t*)data, *data_size));
+    }
 
```

Comment:
The SW convention the eeid enclave loader is suppoed to follow is to copy all fields of the base image `SIGSTRUCT` into the "dynamically" signed `SIGSTRUCT`, exepect `ENCLAVEHASH` and signing related fields. To verify that the eeid enclave loader indeed followed the SW convention, the verifier should perform checks between all relevant fileds in the base image `SIGSTRUCT` and the SGX `QUOTE` fileds, exactly as what the CPU would do. 
- reported_attribute(same as SECS.ATTRIBUTE) & baseSIGSTRUCT.ATTRIBUTEMASK = baseSIGSTRUCT.ATTRIBUTES & baseSIGSTRUCT.ATTRIBUTEMASK,
- reported_misc_select(same as SECS.MISCSELECT) & baseSIGSTRUCT.MISCMASK = baseSIGSTRUCT.MISCSELECT & baseSIGSTRUCT.MISCMASK,
- reported_product_id = baseSIGSTRUCT.ISVPRODID,
- reported_security_version = baseSIGSTRUCT.ISVSVN

If we don't prevent SGX2.0 enclave from being launched as EEID enclave in the future, the check should also inlcude ISVFAMILYID, ISVEXTPRODID

Reply:
Yes, that's exactly what the loader does. The verifier then performs most of those checks here: https://github.com/openenclave/openenclave/blob/c75ff164fda058f914782279dc12d4282223943a/common/sgx/eeid.c#L580 

The attributes change during enclave creation and the enclave doesn't have access to a pre-creation copy of the sigstruct. I'll have to run a few more checks to see if the masks actually do line up as you suggest.

Reply:
I add more masked checks which do pass the tests:  https://github.com/openenclave/openenclave/pull/3236/commits/abfdd68b050e70b3d86a960195f4899a14852d12

Note the negations on the masks; is that the correct masking?

Reply:
Something's off here. The Intel manual also says `&` without `~`, but then the default OE attribute mask doesn't make sense to me. Is the default right?  https://github.com/openenclave/openenclave/blob/05d109a5706fe4c439f64cd68ffee6e2430b9542/include/openenclave/bits/sgx/sgxtypes.h#L136

Reply:
Got it now. It's really just about the "initialized" flag. The base image sigstruct was never created or initialized, so it's missing that flag. I updated the code accordingly.

Reply:
Should not use ~. If a bit in the mask is 0, the loader is free to set the corresponding bit in SECS. If a bit in the mask is 1, the loader must set the corresponding bit in SECS as same as the crespsonding bit in SIGSTRUCT.ATTRIBUTES (or MISCSELECT). So the default SGX_SIGSTRUCT_MISCSELECT=0x00000000,SGX_SIGSTRUCT_MISCMASK=0xffffffff mean the load must set SECS.MISCSELECT=0x00000000 (no opt-in for any MISCSELECT feature). The default of SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS=0XfffffffffffffffbULL, SGX_SIGSTRUCT_ATTRIBUTEMASK_XFRM=0x0000000000000000ULL means the loader is free to enable/disable extended feature set in the enclave, but must make sure most bits in SECS.ATTRIBUTES.FLAGS matches SIGSTRUCT.ATTRIBUTES.FLAGS. I have some doubt about allowing the loader to enable MPX in enclave. I will analyze it later.

Reply:
Outdated. I haven't thought about MPX yet, but in general I think it makes sense to review those defaults. The rest works since last week, with the minor annoyance of having to add the init-bit for the base image.